python/ml4ir/applications/ranking/tests/test_lr_schedule.py

```diff
@@ -0,0 +1,223 @@
+from ml4ir.base.model.scoring.scoring_model import ScorerBase, RelevanceScorer
+from ml4ir.base.model.scoring.interaction_model import InteractionModel, UnivariateInteractionModel
+from ml4ir.base.model.optimizers.optimizer import get_optimizer
+from ml4ir.applications.ranking.model.ranking_model import RankingModel
+from ml4ir.applications.ranking.config.keys import LossKey
+from ml4ir.applications.ranking.config.keys import ScoringTypeKey
+from ml4ir.applications.ranking.model.losses import loss_factory
+import yaml
+from ml4ir.base.features.feature_config import ExampleFeatureConfig, SequenceExampleFeatureConfig
+import tensorflow as tf
+from tensorflow import keras
+import numpy as np
+from tensorflow.keras.optimizers.schedules import ExponentialDecay
+
+warnings.filterwarnings("ignore")
+
+INPUT_DIR = "ml4ir/applications/ranking/tests/data/"
+MODEL_CONFIG = "ml4ir/applications/ranking/tests/data/configs/model_config_cyclic_lr.yaml"
+KEEP_ADDITIONAL_INFO = 0
+NON_ZERO_FEATURES_ONLY = 0
+
+class LrCallback(keras.callbacks.Callback):
+    """Defining a callback to keep track of changing learning rate during training."""
+
+    def __init__(self):
+        self.lr_list = []
+
+    def on_train_batch_begin(self, batch, logs=None):
+        self.lr_list.append(self.model.optimizer._decayed_lr(tf.float32).numpy())
+
+    def get_lr_list(self):
+        return self.lr_list
+
+class TestLrSchedules(unittest.TestCase):
+    """Testing different learning rate schedules"""
+
+    def setUp(self):
+        self.feature_config_yaml_convert_to_clicks = INPUT_DIR + \
+            'ranklib/feature_config_convert_to_clicks.yaml'
+        self.model_config_file = MODEL_CONFIG
+
+    def compare_lr_values(self, scheduler, expected_values):
+        """Expects a learning rate scheduler `scheduler` and
+        a list of values. Compares ...."""
+        actual_values = [scheduler[i].numpy() for i in range(len(expected_values))]
+        return np.all(np.isclose(actual_values, expected_values, rtol=0.001))
+
+    def parse_config(self, tfrecord_type: str, feature_config, io) -> SequenceExampleFeatureConfig:
+        if feature_config.endswith(".yaml"):
+            feature_config = io.read_yaml(feature_config)
+        else:
+            feature_config = yaml.safe_load(feature_config)
+
+        return SequenceExampleFeatureConfig(feature_config, None)
+
+    def test_cyclic_tri_lr(self):
+        """Test a triangular cyclic learning rate"""
+        gold = [0.1,0.28000003,0.45999995,0.64000005,0.81999993,1.0,0.81999993,0.64000005,0.45999995,0.28000003,
+                0.1,0.28000003,0.46000007,0.6399999,0.81999993,1.0,0.81999993,0.6399999,0.46000007,0.28000003,
+                0.1,0.27999982,0.46000007,0.6399999,0.8200002]
+        lrs = cyclic_learning_rate.TriangularCyclicalLearningRate(
+            initial_learning_rate=0.1,
+            maximal_learning_rate=1.0,
+            step_size=5,
+        )
+        scheduler = [lrs(i) for i in range(25)]
```

Comment:
this repeat in every test. You can move it in the `self.compare_lr_values`. If you do, you don't need the 25 (hard-coded now) you can infer it from the `len(gold)` directly.

Reply:
Actually, I recommended we move all these values to a separate gold file to keep the code readable.

Reply:
what you suggest is even better. A JSON or something that holds the expected values of its optimizer.